src/distribution/gamma.rs

```diff
@@ -1,9 +1,78 @@
-use crate::distribution::{Continuous, ContinuousCDF};
+use crate::distribution::{Continuous, ContinuousCDF, ParametrizationError as ParamError};
 use crate::function::gamma;
 use crate::prec;
 use crate::statistics::*;
-use crate::{Result, StatsError};
 use rand::Rng;
+use thiserror::Error;
+
+#[derive(Clone, PartialEq, Debug, Error)]
+pub enum GammaError {
+    #[error("shape must be finite, positive, and not nan")]
+    InvalidShape(#[source] ParamError<f64>),
+    #[error("shape must be finite, positive, and not nan")]
+    InvalidRate(#[source] ParamError<f64>),
+    #[error("rate of {0} is degenerate")]
+    DegenerateRate(f64),
+    #[error("shape of {0} is degenerate")]
+    DegenerateShape(f64),
+}
+
+impl From<super::negative_binomial::NegativeBinomialError> for GammaError {
+    fn from(value: super::negative_binomial::NegativeBinomialError) -> Self {
+        use super::negative_binomial::NegativeBinomialError::*;
+        match value {
+            InvalidMean(e) => Self::InvalidShape(e),
+            InvalidProbability(p) => {
+                if p.is_nan() {
+                    Self::InvalidRate(ParamError::ExpectedNotNan)
+                } else {
+                    Self::InvalidRate(ParamError::ExpectedPositive(p / (1.0 - p)))
+                }
+            }
+            InvalidSuccessCount(e) => Self::InvalidRate(e.into()),
+            DegenerateMean(m) => Self::DegenerateShape(m),
+            DegenerateProbability(p) => Self::DegenerateRate(p / (1.0 - p)),
+            DegenerateSuccessCount => Self::DegenerateRate(0.0),
+        }
+    }
+}
+
+/// holds a valid parametrization of the gamma distribution in shape and rate.
+pub struct Parameters {
+    shape: f64,
+    rate: f64,
+}
```

Comment:
Can you explain why you added this abstraction? You could also check the parameters inside `Gamma::new` and return a suitable error from there, right?

Reply:
Mostly just playing around with having input checking in another type to support the use of infallible `new` methods. 

But more interestingly perhaps it _could_ be a way to get alternative parametrizations #198 as well as `new_from_moments` to do something like,

```rs
let m = Gamma::new(shape, rate).into_moments();
let p = Poisson::new_from_moments(m);
```

But I've lots of thoughts floating around about that none motivated by something I've actually needed thus far. As a solution to #198 could be adding just one more method.